Realm/ObjectStore/impl/async_query.cpp

```diff
@@ -0,0 +1,265 @@
+    m_realm = nullptr;
+}
+
+size_t AsyncQuery::add_callback(std::function<void (std::exception_ptr)> callback)
+{
+    m_realm->verify_thread();
+
+    auto next_token = [=] {
+        size_t token = 0;
+        for (auto& callback : m_callbacks) {
+            if (token <= callback.token) {
+                token = callback.token + 1;
+            }
+        }
+        return token;
+    };
+
+    std::lock_guard<std::mutex> lock(m_callback_mutex);
+    auto token = next_token();
+    m_callbacks.push_back({std::move(callback), token, -1ULL});
+    if (m_callback_index == npos) { // Don't need to wake up if we're already sending notifications
+        Realm::Internal::get_coordinator(*m_realm).send_commit_notifications();
+        m_have_callbacks = true;
+    }
+    return token;
+}
+
+void AsyncQuery::remove_callback(size_t token)
+{
+    Callback old;
+    {
+        std::lock_guard<std::mutex> lock(m_callback_mutex);
+        REALM_ASSERT(m_error || m_callbacks.size() > 0);
+
+        auto it = find_if(begin(m_callbacks), end(m_callbacks),
+                          [=](const auto& c) { return c.token == token; });
+        // We should only fail to find the callback if it was removed due to an error
+        REALM_ASSERT(m_error || it != end(m_callbacks));
+        if (it == end(m_callbacks)) {
+            return;
+        }
+
+        size_t idx = distance(begin(m_callbacks), it);
+        if (m_callback_index != npos && m_callback_index >= idx) {
+            --m_callback_index;
+        }
+
+        old = std::move(*it);
+        m_callbacks.erase(it);
+    }
+}
+
+void AsyncQuery::unregister() noexcept
+{
+    std::lock_guard<std::mutex> lock(m_target_mutex);
+    m_target_results = nullptr;
+    m_realm = nullptr;
+    m_unregistered = true;
+}
+
+void AsyncQuery::release_query() noexcept
+{
+    {
+        std::lock_guard<std::mutex> lock(m_target_mutex);
+        REALM_ASSERT(m_unregistered);
+    }
+
+    m_query = nullptr;
+}
+
+bool AsyncQuery::is_alive() const noexcept
+{
+    std::lock_guard<std::mutex> lock(m_target_mutex);
+    return !m_unregistered;
+}
+
+void AsyncQuery::run()
+{
+    REALM_ASSERT(m_sg);
+
+    // This function must not touch any members touched in deliver(), as they
+    // may be called concurrently (as it'd be pretty bad for a running query to
+    // block the main thread trying to pick up the previous results)
+    {
+        std::lock_guard<std::mutex> target_lock(m_target_mutex);
+        // Don't run the query if the results aren't actually going to be used
+        if (!m_target_results || (!m_have_callbacks && !m_target_results->wants_background_updates())) {
+            return;
+        }
+    }
+
+    REALM_ASSERT(!m_tv.is_attached());
+
+    // If we've run previously, check if we need to rerun
+    if (m_initial_run_complete) {
+        // Make an empty tableview from the query to get the table version, since
+        // Query doesn't expose it
+        m_tv = m_query->find_all(0, 0, 0);
```

Comment:
Why does this assign to `m_tv` rather than a local temporary? From what I can see all paths that hit this line proceed to assign a new value to `m_tv`.